skll/data/writers.py

```diff
@@ -221,145 +227,125 @@ def _write_line(self, id_, label_, feat_dict, output_file):
+
+        # create a list of tuples with (column names, column indexes)
+        # so that we can correctly extract the appropriate columns
+        columns = sorted([(col_name, col_idx) for col_name, col_idx
+                          in feature_set.vectorizer.vocabulary_.items()
+                          if (col_name in filter_features or
+                              col_name.split('=', 1)[0] in filter_features)],
+                         key=lambda x: x[1])
+
+        # then, split the names and indexes into separate lists
+        column_names, column_indexes = zip(*columns)
+        return list(column_names), list(column_indexes)
+
+    def _create_dataframe(self, feature_set, filter_features=None):
+        """
+        Create the data frame from the feature set.feature
 
         Parameters
         ----------
-        id_ : str
-            The ID for the current instance.
-        label_ : str
-            The label for the current instance.
-        feat_dict : dict
-            The feature dictionary for the current instance.
-        output_file : file buffer
-            The file being written to.
+        feature_set : skll.FeatureSet
+            The ``FeatureSet`` instance being written to a file.
+        filter_features : set of str or None, optional
+            If only writing a subset of the features in the
+            FeatureSet to ``output_file``, these are the
+            features to include in this file.
+            Defaults to None.
+
+        Returns
+        -------
+        df : pd.DataFrame
+            The data frame constructed from
+            the feature set.
 
         Raises
         ------
         ValueError
-            If class column name is already use as a feature
-        ValueError
-            If ID column name is already used as a feature.
+            If ID column is already used as feature.
+            If label column is already used as feature.
         """
-        # Add class column to feat_dict (unless this is unlabelled data)
-        if self.label_col not in feat_dict:
-            if self.feat_set.has_labels:
-                feat_dict[self.label_col] = label_
-        else:
-            raise ValueError(('Class column name "{}" already used as feature '
-                              'name.').format(self.label_col))
-        # Add id column to feat_dict if id is provided
-        if self.id_col not in feat_dict:
-            feat_dict[self.id_col] = id_
-        else:
+        # if there is no filtering, then just keep all the names
+        (column_names,
+         column_idxs) = self._get_column_names_and_indexes(feature_set,
+                                                           filter_features)
+
+        # create the data frame from the feature set;
+        # then, select only the columns that we want,
+        # and give the columns their correct names
+        df = pd.DataFrame(feature_set.features.todense())
```

Comment:
Hmm, is there no way around converting things to `dense` and we are making a copy as well. Should we check if we are trading off a lot of memory again just like when reading?

Reply:
I'm not sure of another way to convert a sparse array to a `DataFrame`, other than calling `toarray()` or `todense()`. Is `toarray()` more efficient? Or is there some alternative I'm not thinking of?